hilti/toolchain/include/ast/scope.h

```diff
@@ -12,33 +12,51 @@
 #include <utility>
 #include <vector>
 
-#include <hilti/ast/node-ref.h>
-#include <hilti/base/intrusive-ptr.h>
+#include <hilti/ast/forward.h>
 
 namespace hilti {
 
 class Declaration;
 class ID;
 
 /**
- * Identifier scope. A scope maps identifiers to AST nodes (more precisely: to
- * references to AST nodes). An identifier can be mapped to more than one node.
+ * Scope mapping a set of identifiers to target declarations. An identifier can
+ * be mapped to more than one target.
  */
-class Scope : public intrusive_ptr::ManagedObject {
+class Scope {
 public:
     Scope() = default;
     ~Scope() = default;
 
-    void insert(NodeRef&& n);
-    void insert(const ID& id, NodeRef&& n);
+    /**
+     * Inserts a declaration into the scope.
+     *
+     * @param d declaration to insert
+     */
+    void insert(DeclarationPtr&& d);
```

Comment:
Is there a benefit to only accepting r-values here, e.g., since accepting a `DeclarationPtr` wouldn't trigger warnings about needless copies in callers?

Here and below.

Reply:
Left-over.